src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCheckedParameterTests.cs

```diff
@@ -1264,5 +1300,417 @@ .maxstack  1
+    IL_0001:  brtrue.s   IL_000e
+    IL_0003:  ldstr      ""s""
+    IL_0008:  newobj     ""System.ArgumentNullException..ctor(string)""
+    IL_000d:  throw
+    IL_000e:  ldc.i4.0
+    IL_000f:  newobj     ""C.<GetChars>d__1..ctor(int)""
+    IL_0014:  ret
+}");
+        }
+
+        [Fact]
+        public void TestNullCheckedOutString()
+        {
+            var source = @"
+class C
+{
+    public static void Main() { }
+    public void M(out string x!)
+    {
+        x = ""hello world"";
+    }
+}";
+            var compilation = CreateCompilation(source);
+            compilation.VerifyDiagnostics(
+                    // (5,31): error CS8720: By-reference parameter 'out string x!' cannot be null-checked.
+                    //     public void M(out string x!)
+                    Diagnostic(ErrorCode.ERR_NullCheckingOnByRefParameter, "!").WithArguments("out string x!").WithLocation(5, 31));
+        }
+
+        [Fact]
+        public void TestNullCheckedRefString()
+        {
+            var source = @"
+class C
+{
+    public static void Main() { }
+    public void M(ref string x!)
+    {
+        x = ""hello world"";
+    }
+}";
+            var compilation = CreateCompilation(source);
+            compilation.VerifyDiagnostics(
+                    // (5,31): error CS8720: By-reference parameter 'ref string x!' cannot be null-checked.
+                    //     public void M(ref string x!)
+                    Diagnostic(ErrorCode.ERR_NullCheckingOnByRefParameter, "!").WithArguments("ref string x!").WithLocation(5, 31));
+        }
+
+        [Fact]
+        public void TestNullCheckedInString()
+        {
+            var source = @"
+class C
+{
+    public static void Main() { }
+    public void M(in string x!) { }
+}";
+            var compilation = CreateCompilation(source);
+            compilation.VerifyDiagnostics( 
+                    // (5,30): error CS8720: By-reference parameter 'in string x!' cannot be null-checked.
+                    //     public void M(in string x!)
+                    Diagnostic(ErrorCode.ERR_NullCheckingOnByRefParameter, "!").WithArguments("in string x!").WithLocation(5, 30));
+        }
+
+        [Fact]
+        public void TestNullCheckedParams()
+        {
+            var source = @"
+class C
+{
+    public static void Main() { }
+    public void M(params int[] number!) {}
+}";
+            CompileAndVerify(source).VerifyIL("C.M(params int[])", @"
+{
+    // Code size       15 (0xf)
+    .maxstack  1
+    IL_0000:  ldarg.1
+    IL_0001:  brtrue.s   IL_000e
+    IL_0003:  ldstr      ""number""
+    IL_0008:  newobj     ""System.ArgumentNullException..ctor(string)""
+    IL_000d:  throw
+    IL_000e:  ret
+}");
+        }
+
+        [Fact]
+        public void TestNullCheckedGenericWithDefault()
+        {
+            var source = @"
+class C
+{
+    static void M1<T>(T t! = default) { }
+}";
+            CompileAndVerify(source, options: TestOptions.DebugDll).VerifyIL("C.M1<T>(T)", @"
```

Comment:
>CompileAndVerify [](start = 12, length = 16)

Please use `comp.VerifyDiagnostics()` in this test and two tests below because `CompileAndVerify()` will ignore warnings and we're testing for `WRN_NullCheckedHasDefaultNull` in these three tests.